### PR TITLE
Rooms repo

### PR DESCRIPTION
**Pull request by sychova**

### CODE REVIEW

File: db/migrations/20220106131805_create_rooms_users_connection.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
+exports.up = async (knex) => {
+    await knex.schema.createTable('rooms_users', (t) => {
```
Comment (alexandr-senyuk) [inline db/migrations/20220106131805_create_rooms_users_connection.js:2]: add `room_id&user_id` uniqueness constraint

File: src/repos/roomRepo.js [hunk 2]
```diff
@@ -9,52 +10,50 @@ class RoomRepo extends BaseRepo {
...
 
-    removeUserFromRoom(room, user) {
-        const record = this.gateway.table.get(room.title)
+    async removeUserFromRoom(room, user) {
+        const record = await this.tryFindByTitle(room.title)
         if (!record) this.throwNotFound()
 
-        record.users.delete(user.id)
-        this.gateway.table.set(room.title, record)
+        await knex('rooms_users').where('user_id', user.id).del()
```
Comment (alexandr-senyuk) [inline src/repos/roomRepo.js]:
```suggestion
        await this.gateway('rooms_users').where('user_id', user.id).del()
```

Comment (alexandr-senyuk) [reply src/repos/roomRepo.js]: in the whole file

File: src/repos/roomRepo.js [hunk 3]
```diff
@@ -1,6 +1,7 @@
 const BaseRepo = require('./baseRepo')
 const { Room } = require('../entities')
 const { NotFound } = require('./common')
+const knex = require('../utils/knex')
```
Comment (alexandr-senyuk) [inline src/repos/roomRepo.js]: get rid

File: src/repos/roomRepo.js [hunk 4]
```diff
@@ -1,6 +1,7 @@
...
 
 let seq = 1
```
Comment (alexandr-senyuk) [inline src/repos/roomRepo.js]: get rid, please. my bad

File: src/repos/roomRepo.js [hunk 5]
```diff
@@ -9,52 +10,50 @@ class RoomRepo extends BaseRepo {
...
 
-    get all() {
-        const result = []
-
-        this.gateway.keys.forEach((key) =>
-            result.push(this.gateway.table.get(key))
-        )
-        return result.filter(Boolean)
+    async all() {
+        return await this.query.select()
```
Comment (alexandr-senyuk) [inline src/repos/roomRepo.js]: redundant async-await here

File: src/repos/roomRepo.js [hunk 6]
```diff
@@ -9,52 +10,50 @@ class RoomRepo extends BaseRepo {
...
     }
 
-    getUsersForRoom(room) {
-        return this.userRepo.getByIds([...room.users])
+    async getUsersForRoom(room) {
+        const record = await this.tryFindByTitle(room.title)
+        const usersInRoom = await knex('rooms_users')
+            .where({ room_id: record.attr.id })
+            .select('user_id')
+            .then((users) => users.map((r) => r.user_id))
```
Comment (alexandr-senyuk) [inline src/repos/roomRepo.js]: use `pluck`
https://knexjs.org/#Builder-pluck

File: src/repos/roomRepo.js [hunk 7]
```diff
@@ -9,52 +10,50 @@ class RoomRepo extends BaseRepo {
...
-        )
-        return result.filter(Boolean)
+    async all() {
+        return await this.query.select()
     }
 
-    getUsersForRoom(room) {
-        return this.userRepo.getByIds([...room.users])
+    async getUsersForRoom(room) {
+        const record = await this.tryFindByTitle(room.title)
```
Comment (alexandr-senyuk) [inline src/repos/roomRepo.js]: prefer using `id` for search if you have one

**Review by alexandr-senyuk — CHANGES_REQUESTED** [review]

File: db/migrations/20220106131805_create_rooms_users_connection.js (hunk 1 above)
Comment (sychova) [reply db/migrations/20220106131805_create_rooms_users_connection.js:2]: Fixed.

File: src/repos/roomRepo.js (hunk 2 above)
Comment (sychova) [reply src/repos/roomRepo.js]: Fixed.

File: src/repos/roomRepo.js (hunk 3 above)
Comment (sychova) [reply src/repos/roomRepo.js]: Fixed.

File: src/repos/roomRepo.js (hunk 4 above)
Comment (sychova) [reply src/repos/roomRepo.js]: Fixed.

File: src/repos/roomRepo.js (hunk 6 above)
Comment (sychova) [reply src/repos/roomRepo.js]: Fixed.

File: src/repos/roomRepo.js (hunk 5 above)
Comment (sychova) [reply src/repos/roomRepo.js]: Fixed.

File: src/repos/roomRepo.js (hunk 7 above)
Comment (sychova) [reply src/repos/roomRepo.js]: Fixed.

File: src/repos/roomRepo.js [hunk 8]
```diff
@@ -2,59 +2,61 @@ const BaseRepo = require('./baseRepo')
...
-        )
-        return result.filter(Boolean)
+    all() {
+        return this.query.select()
     }
 
-    getUsersForRoom(room) {
-        return this.userRepo.getByIds([...room.users])
+    async getUsersForRoom(room) {
+        const record = await this.tryFindById(room.attr.id)
```
Comment (alexandr-senyuk) [inline src/repos/roomRepo.js]: 1) update `Room#id` getter to access it directly
2) better to define `findById` with `this.mapOrNotFound` inside

Comment (sychova) [reply src/repos/roomRepo.js]: Fixed.

Comment (alexandr-senyuk) [reply src/repos/roomRepo.js]: `2. better to define findById` is not 
please, rename the method and that's it

File: src/repos/roomRepo.js [hunk 9]
```diff
@@ -2,59 +2,61 @@ const BaseRepo = require('./baseRepo')
...
-        return result.filter(Boolean)
+    all() {
+        return this.query.select()
     }
 
-    getUsersForRoom(room) {
-        return this.userRepo.getByIds([...room.users])
+    async getUsersForRoom(room) {
+        const record = await this.tryFindById(room.id)
+        if (!record) this.throwNotFound()
```
Comment (alexandr-senyuk) [inline src/repos/roomRepo.js]: you don't need this line anymore and next empty line as well

File: src/repos/roomRepo.js (hunk 8 above)
Comment (sychova) [reply src/repos/roomRepo.js]: Fixed.

File: src/repos/roomRepo.js (hunk 9 above)
Comment (sychova) [reply src/repos/roomRepo.js]: Fixed.

File: src/repos/index.js [hunk 10]
```diff
@@ -3,10 +3,7 @@ const UserRepo = require('./userRepo')
 const RoomRepo = require('./roomRepo')
 
 const userRepo = new UserRepo(knex)
-const roomRepo = new RoomRepo(
-    { keys: new Set(), table: new Map() },
-    { userRepo }
-)
+const roomRepo = new RoomRepo(knex, { userRepo })
```
Comment (alexandr-senyuk) [inline src/repos/index.js]: `userRepo` is no need

File: src/repos/userRepo.js [hunk 11]
```diff
@@ -25,9 +25,11 @@ class UserRepo extends BaseRepo {
         return this.mapOrNull(record)
     }
 
-    async getByIds(ids) {
-        const records = await this.query.whereIn('id', ids)
-        return this.map(records)
+    async getUsersForRoom(room) {
+        const usersInRoom = await this.gateway('rooms_users')
+            .join('users', 'users.id', '=', 'rooms_users.user_id')
```
Comment (alexandr-senyuk) [inline src/repos/userRepo.js]: equality operator is default 
https://knexjs.org/#Builder-join

Comment (sychova) [reply src/repos/userRepo.js]: Fixed.

File: src/repos/index.js (hunk 10 above)
Comment (sychova) [reply src/repos/index.js]: Fixed.

**Pull request merged**